urlac47192b66_ggrandes_bouncer_tgz-pJ8-org_javastack_bouncer_ClusterTestJ8/src/org/javastack/bouncer/MuxClient.java

```diff
@@ -0,0 +1,490 @@
+package org.javastack.bouncer;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.EOFException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.ConnectException;
+import java.net.InetAddress;
+import java.net.Socket;
+import java.net.SocketTimeoutException;
+import java.security.GeneralSecurityException;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.TimeUnit;
+
+import javax.net.ssl.SSLSocket;
+
+// ============================================ Mux Client
+
+// MuxClient (MUX=OUT) Local=RAW, Remote=MUX
+class MuxClient {
+	final MuxClientMessageRouter router = new MuxClientMessageRouter();
+	final HashMap<Integer, MuxClientLocal> mapLocals = new HashMap<Integer, MuxClientLocal>();
+	final ServerContext context;
+	final Map<Integer, OutboundAddress> left;
+	final OutboundAddress right;
+
+	MuxClientRemote remote;
+
+	MuxClient(final ServerContext context, final OutboundAddress left, final OutboundAddress right) {
+		this.context = context;
+		this.left = (left == null ? new HashMap<Integer, OutboundAddress>() : //
+				Collections.singletonMap(Integer.valueOf(0), left));
+		this.right = right;
+	}
+
+	void addLeft(final OutboundAddress left) {
+		this.left.put(left.getOpts().getTunID(), left);
+	}
+
+	void openRemote() throws IOException { // Entry Point
+		Log.info(this.getClass().getSimpleName() + "::openRemote " + right);
+		remote = new MuxClientRemote(right);
+		remote.setRouter(router);
+		context.addShutdownable(remote);
+		context.submitTask(remote, "MuxOutRight[" + right + "]", ClientId.newId());
+	}
+
+	void openLocal(final int id, final int idEndPoint, final InetAddress srcAddress) {
+		Log.info(this.getClass().getSimpleName() + "::openLocal id=" + id + " srcAddr="
+				+ String.valueOf(srcAddress));
+		final OutboundAddress left = this.left.get(Integer.valueOf(idEndPoint));
+		final MuxClientLocal local = new MuxClientLocal(left);
+		local.setId(id);
+		local.setSticky(srcAddress);
+		local.setRouter(router);
+		synchronized (mapLocals) {
+			mapLocals.put(id, local);
+		}
+		context.submitTask(local, "MuxOutLeft-Recv[" + id + "|" + right + "]",
+				(((long) id << 48) | ClientId.newId()));
+	}
+
+	void closeLocal(final int id) {
+		// Send FIN
+		try {
+			final MuxPacket mux = context.allocateMuxPacket();
+			mux.fin(id);
+			remote.sendRemote(mux);
+			context.releaseMuxPacket(mux);
+		} catch (Exception ign) {
+		}
+		synchronized (mapLocals) {
+			final MuxClientLocal local = mapLocals.remove(id);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code might contain a resource leak. A resource leak can slow down or crash your system.

**Fix**
Consider closing the following resource in a try-finally block: *local*. Execution paths that do not contain closure statements were detected.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).